### DISCUSSION TITLE
Mobile UI proposal

### DISCUSSION THREAD

**Discussion by ngocdh** [jamulussoftware/jamulus#1766] [UI Related]
We started a discussion about mobile UI in #1765. While I love using the one window interface on desktop, I think mobile apps must have their own design because of small screens (except for tablets where the desktop design works fine).

So here's my POC for the mobile UI, please forgive my poor skills, but you should get the idea. I'm showing the element which I think is the most important and need most attention: the Mixer. Other screens (Connect, Chat, Settings, etc. should look similarly to current design).

![image](https://user-images.githubusercontent.com/6765612/119832785-cd48c100-befe-11eb-8155-fa4a4ec1531f.png)

What do you think?

**Comment by ann0see** [comment]
Yes. That would be a great starting point. Probably you'll need to do some stuff with QML I don't have any experience on QML, but I'm afraid that this involves a lot of restructuring of the source code. However, since you got the sound system on iOS working, I'm sure it's not impossible.

Tagging @jamulussoftware/designers here. They might have some ideas.

**Comment by gilgongo** [comment]
Good that this is getting some attention. Interesting too that you appear to have grouped "my controls" **with** your sider, and not have that slider as it currently is along with all the other participants. This was the subject of some discussion a while ago: does having your own fader in the "main mix" panel confuse things in relation to (for example) what "mute myself" does? A possible desktop re-configuration of that was:

![image](https://user-images.githubusercontent.com/4561747/119847159-a55b5c80-bf02-11eb-8db4-2c33cc14c0f3.png)

But to return to your design: For handsets, the basic requirement for typical usability of interactive elements is that "hit areas" (the area that responds to a touch, as opposed  to the "touch target" which is the area that looks interactive) need to be a minimum of 6mm on centre. Note also that best practice is to favour larger muscle movements for more important thumb interactions, which would mean setting faders sideways perhaps. But it's a bit hard to say at this stage.

**Reply by ngocdh** [reply]
Thanks for your input. I agree controls need to be larger, and have more space between them.

About "My controls" I think "Practice mode" instead of "Mute Myself" might clear the confusion.

**Reply by ngocdh** [reply]
And the other should be "Self-monitor" which is ON by default.

"Practice mode" and "Self-monitor"

**Reply by dcorson-ticino-com** [reply]
From my point of view you are one of the musicians, thus your fader should be with the others, not alone at the top.

Being alone at the top, separated from the others, gives the impression that your fader is an input volume control that makes you louder or softer for everyone.  This is, however, not the case.  Your fader only changes the volume that you hear and is thus exactly analogue to the faders of all the other musicians.

Your input volume control is a hardware function of your device, it is not part of Jamulus (the output volume too for that matter).
I would have, however, nothing against putter your own fader always as the first fader.

**Reply by ngocdh** [reply]
Don, I think your point of view comes from someone who writes code for Jamulus, and/or know Jamulus from the inside out. You know its logic and want to preserve that logic.

But an app should be **User-oriented**. You should look at it as a user. 

As a user, I don't care if my fader has the exact same logic as others. I don't care if people hear me differently if I change my fader (actually I know they also have a faders to adjust my volume). So, the design should serve best the user who is using the app. And for me, the user deserve a dedicated area for all their controls.

**Reply by dcorson-ticino-com** [reply]
>But an app should be User-oriented. You should look at it as a user.

Exactly, I hear the complaints already "I have pushed the fader up all the way and the others still can't hear me any louder"
(Well of course not...)

**Reply by ngocdh** [reply]
See, the fader was among the others and that user still didn't understand its logic.

**Reply by dcorson-ticino-com** [reply]
Because you are not giving then the clue that it is the same.

**Reply by ngocdh** [reply]
Oh sorry I thought you said "heard".

I actually thought about putting the input gain control there. Now that you bring this up, I think we have a good reason to do it :)

**Reply by gilgongo** [reply]
I think, as @ngocdh implies, if someone says, _"I have pushed the fader up all the way and the others still can't hear me any louder"_ then they have a somewhat bigger problem, because that indicates they think they think the faders act for all (eg _"I have pushed **the bassist's** fader up all the way and the others still can't hear **them** any louder"_). 

There was a reason why the input gain wasn't put into the main window, but I can't recall what that was now. I think @hoffie knows. (other than having two faders would of course be even more confusing)

**Reply by ngocdh** [reply]
> There was a reason why the input gain wasn't put into the main window

If I recall correctly, the input gain is meant to be used once at initial setup only, not frequently. But I'd love to use it to reduce my volume in people's mix too, because it's not easy doing it at OS level, and it's just a few lines of code anyway.

**Comment by ngocdh** [comment]
Here's an updated POC with larger controls, and input gain as a dropdown box (like what we have now in Settings), not a fader to avoid the confusion.

![image](https://user-images.githubusercontent.com/6765612/119868993-0a717b00-bf20-11eb-9411-b1c33bb91e2f.png)

**Reply by gilgongo** [reply]
6mm on centre would mean the "Grp | Mute | Solo" stack would be 18mm high,  I think the most you'd be realistically able to get on most screens would be two rows with that layout, even without considering the fact that the pan controls would overlap with the fader at the top quite a lot. 

I think a handset layout for thumb control to avoid interference would need to be more like this at about 375px perhaps:

![image](https://user-images.githubusercontent.com/4561747/119889539-99858f80-bf2e-11eb-96de-388097fe8940.png)

**Reply by ngocdh** [reply]
That's really nice @gilgongo. You are the expert here. What did you use to create that POC?

**Reply by ngocdh** [reply]
I'm thinking that would also mean a complete rewrite of the UI for mobile without reusing existing (UI related) code?

**Reply by ngocdh** [reply]
@gilgongo like this? Sorry I don't speak pixel :(

![image](https://user-images.githubusercontent.com/6765612/119897825-47019e80-bf41-11eb-91ab-d45e357dba09.png)

**Reply by gilgongo** [reply]
Yes, something like that maybe so that the various touch targets are reasonably separated (but I'm sure it needs more thought). I've not actually verified that sketch for pixel accuracy - just estimated. And it would of course imply a complete UI re-write :-(  The tool I used for the mockup is called Axure.

**Comment by ngocdh** [comment]
So here's my 1st attempt to write a mobile app using QML:
https://github.com/ngocdh/jamulus/tree/qml

![image](https://user-images.githubusercontent.com/6765612/120164018-68d97a80-c1fa-11eb-82da-3e9d30c3b216.png)


The App is just a very simple Jamulus client which connects to a server given the address. In the mixerboard, there's only a "Practice Mode" button. Works on iOS, not yet on Android.

**Reply by gilgongo** [reply]
Nice!

BTW what's the current thinking about networking on mobile? There was a time when I thought that because clients would be on WiFi then mobile would only be any good for listening. But is WiFi on mobile in fact OK to play over your experience?

"Practice mode" is currently "Mute myself" in the desktop app, is that right?

**Reply by ngocdh** [reply]
> BTW what's the current thinking about networking on mobile? ... is WiFi on mobile in fact OK to play over your experience?

I have a good 3-band Wifi router. Delay is good when using 5.0Ghz wifi. I have been playing ok on mobile with it.

It also seems to work for @ann0see using Ethernet.

> "Practice mode" is currently "Mute myself" in the desktop app, is that right?

Yes correct.

**Reply by dcorson-ticino-com** [reply]
We have to be sure to separate mobile phone connectivity with GSM 4G or 5G from wifi of any type.
I have been able to use a mobile hot spot using GSM 4G with a wired Ethernet connection for a reasonable (but not really good) connection. Using the wifi of the hot spot, both so-call 4g and 5g, for the last leg instead of the wired cable made the Internet connection unusable for Jamulus, the jitter was way too high added to the GSM 4G jitter.
Thus I think it is not impossible that using mobilephone hardware with GSM 4G or 5G directly, without wifi, could give usable connections.  But I have little hope for any kind of wifi.

**Reply by ngocdh** [reply]
Hi Don, this is my Wifi:

![image](https://user-images.githubusercontent.com/6765612/120220218-b4ad1380-c23c-11eb-9461-bc4190d99ce9.png)

**Reply by ngocdh** [reply]
And my 5G :(

![image](https://user-images.githubusercontent.com/6765612/120221497-cdb6c400-c23e-11eb-8bc3-ae02128a7897.png)

**Reply by ghost** [reply]
5GHz WiFi6 with the router's QoS set for the Jamulus device should be surprisingly good.

**Reply by dcorson-ticino-com** [reply]
My connection with a cable, note that if I use wifi, my jitter is similar to yours.  I expect that with a cabled connection you will shave about 10ms from your overall delay.
![Jitter](https://user-images.githubusercontent.com/64357360/120221649-0fe00580-c23f-11eb-939a-9683c0fe78fd.JPG)

Is that GSM 5G?  How far are you from the sending antenna and how many walls does the signal have to go through?
Maybe GSM 4G is better ?
All these wireless connection, GSM and wifi, are also dependent on the number of other users using the frequency.  Using the hot-spot I have noticed that the speed and delay is much better at 4AM than 4PM, just the number of people using their phones at different times of the day.

**Reply by dcorson-ticino-com** [reply]
> 5GHz WiFi6 with the router's QoS set for the Jamulus device should be surprisingly good.

Looking forward to hear more about it.

**Reply by ghost** [reply]
I can't test because the laptop is not WiFi6 and is only 2.4GHz. I do have a new and low cost TP-Link Archer AX20 WiFi 6 Router that significantly improved my ethernet (laptop: speedtest 25 -> 30 megabits/s) and wireless (iPhone: WiFi 6 @ 5GHz)

**Comment by mulyaj** [comment]
If I recall correctly, qml has more flexibility in the way of custom components. Since a rewrite of the code might happen, I'm curious if we could take an [atomic design](https://bradfrost.com/blog/post/atomic-web-design/) approach and build around a design system. It would allow us to get a more consistent aesthetic as well as address scalability in the design (additional features, new contributors, etc.)

**Reply by ngocdh** [reply]
Good idea. 
When I was writing the POC, I didn't have (still don't) have a good understanding of the qml model, just assembling things so that it worked.

It would be great if someone could make that design happen.

**Reply by ann0see** [reply]
@mulyaj  sorry for deleting your comment. I'm a bit worried about posting the clear link here. @ngocdh could you please E-Mail team@jamulus.io to get a Discord/Matrix invite and discuss everything with @mulyaj?
